### PR TITLE
[CPDEV-94365] Add support for RHEL9 family,  SELinux

### PR DESCRIPTION
**Pull request by krang404**
### Description
Selinux 3.5 introduces real support for security contexts. Therefore, installing on the latest versions of rhel-like OS will always fail. That happens because in scripts that send configuration files to a remote host, the files are moved from /tmp directory preserving the source context.

### Solution
* Support operating systems based on RHEL 9 - RHEL 9, Oracle Linux 9, Rocky Linux 9, CentOS 9 Stream. 
* Move files from /tmp setting the default SELinux context of the target location, using `mv -Z` option.

### Tests

Procedure installation, backup and restore tested on RHEL 9, Oracle Linux 9.2, Rocky Linux 9.2 and CentOS 9 Stream

### Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Integration CI passed
- [x] Unit tests. If Yes list of new/changed tests with brief description
- [x] There is no merge conflicts

[Integration Tests](https://github.com/krang404/KubeMarine/actions/workflows/integration_tests.yaml)

#### Unit tests
- def test_specific_os_family_cri_association_upgrade_required
Add rhel9 family
- def _packages_for_cri_os_family
Add rhel9 family os

### CODE REVIEW

File: kubemarine/cri/containerd.py [hunk 1]
```diff
@@ -24,6 +24,17 @@
 from kubemarine.core import utils
 from kubemarine.core.group import NodeGroup, RunnersGroupResult, CollectorCallback
 
+def add_yum_repo_docker(group: NodeGroup)-> RunnersGroupResult:
```
Comment (koryaga) [inline kubemarine/cri/containerd.py:27]: Is specifying repo and extra package in cluster.yaml not working ?
https://github.com/Netcracker/KubeMarine/blob/main/examples/cluster.yaml/allinone-cluster-ol7.yaml#L33
https://github.com/Netcracker/KubeMarine/blob/main/examples/cluster.yaml/allinone-cluster-ol7.yaml#L56

**Comment by krang404** [comment]
In some cases keepalived.service will not start. It nappens, cause **enable service --now**  enable and start service with default config wich contain name eth0. In some virtual machines (like in cloud) default interface named differently. And in new version of systemd start with wrong config will be crushed. I changed logic to install keepalived. In last commit function **install**  in keepalived.py will do install and patch. Service enabled and start in function **configure** .

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (krang404) [reply kubemarine/cri/containerd.py:27]: There are several reasons why I did it this way. Let me tell you about the most important ones.
Firstly, the docker and containerd packages are required to install kubernetes on any OS. In Debian-like distributions, these packages are available immediately, while in systems based on RHEL, it is necessary to connect an additional repository. Therefore, it is logical to transfer the mandatory step when deploying a kubernetes cluster to flow kubemarine. This will make it easier for users to configure the deployment. Isn't that the purpose of any automation? In addition, there is the podman. In ubuntu 20.04, for example, the repository with it needs to be connected separately. I or anyone else in the future will be able to add this function to the cri.add_repo.
Secondly, it is unsafe to connect repositories without GPG authentication. And if you enable authentication in the configuration file that you propose as a template, you will always get an error in installation packages, because the add_repo function only adds the repository to the predefined.repo file without GPG-key. 
Therefore, I consider it justified to introduce a separate step to add the required third-party repositories.

Comment (ilia1243) [reply kubemarine/cri/containerd.py:27]: Hello @krang404,
To clarify, the main goal of the application is to install Kubernetes cluster in production environment. Thus, the main case is when public internet resources are unavailable. The following solutions are developed for this: [Installation without Internet Resources](https://github.com/Netcracker/KubeMarine/blob/main/documentation/Installation.md#installation-without-internet-resource), [package_manager](https://github.com/Netcracker/KubeMarine/blob/main/documentation/Installation.md#package_manager). They allow to specify other custom sources of packages, images, not only the missed docker repo.

This PR is about installation in public environment. It is also supported but should not contradict to the private mode. You add public repo in any mode that looks at least strange.

Returning to your comment. In other words, the approach suggested by @koryaga works in unsafe mode, doesn't it? When repositories are specified in `cluster.yaml`, the installation is automated, and the same approach is applicable for ubuntu 20.04. The flaw related to authentication can potentially be fixed by extending of the [PackageManager.add_repo()](https://github.com/Netcracker/KubeMarine/blob/0.19.0/kubemarine/packages.py#L512) procedure (and corresponding installation task `prepare.package_manager.configure`) in generic way for both public and private environments. What we can think of is again possible extending of the `prepare.package_manager.configure` task to install some repositories by default in some (to be developed) cases.

File: kubemarine/keepalived.py [hunk 2]
```diff
@@ -276,6 +277,6 @@ def configure(group: NodeGroup) -> RunnersGroupResult:
 
     log.debug(group.sudo('ls -la %s' % package_associations['config_location']))
 
-    restart(group)
+    enable(group)
```
Comment (ilia1243) [inline kubemarine/keepalived.py:280]: It should be possible to change keepalived configuration in `cluster.yaml` and run `install --tasks prepare.loadbalancer.keepalived.configure` to configure and apply the changes. I think restart of keepalived should be kept.

File: kubemarine/resources/configurations/globals.yaml [hunk 3]
```diff
@@ -331,7 +351,7 @@ compatibility_map:
       - os_family: 'rhel8'
         versions:
           - '8.6'
-          - '8.7'
+          - '8.8'
```
Comment (ilia1243) [inline kubemarine/resources/configurations/globals.yaml:354]: Probably add 8.8, not replace already supported 8.7?

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (krang404) [reply kubemarine/cri/containerd.py:27]: Hello @ilia1243 First of all, thank you for noticing that the lack of access to the docker repository is not handled in any way. I fixed it. But I also want to tell you that in an environment where there is no SNAT on the Internet, there will also be no access to the repositories from the example above and this will also require a configuration of private resources to install. In such circumstances, it is enough for the user to do ``--exclude prepare.cri.add_repo``. Secondly, I think another way to add the necessary repositories corresponds to the Linux approach, where everything can be done in several ways. And while the alternative you're talking about is still in the thought experiment phase, this fix will help people who install a kubernetes cluster on nodes with RHEL-like operating systems.

File: kubemarine/keepalived.py (hunk 2 above)
Comment (krang404) [reply kubemarine/keepalived.py]: If you mean `install --tasks deploy.loadbalancer.keepalived.configure`, then you are right. I did both restart and enable. This will not break anything and will allow the unit to start correctly.

File: kubemarine/resources/configurations/globals.yaml (hunk 3 above)
Comment (krang404) [reply kubemarine/resources/configurations/globals.yaml]: add 8.7 version

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (ilia1243) [reply kubemarine/cri/containerd.py:27]: > lack of access to the docker repository is not handled in any way.

It is handled through `services.packages.package_manager.repositories` section of the configuration though without GPG-key. Please confirm that the installation works in this case.

> it is enough for the user to do `--exclude prepare.cri.add_repo`

Unfortunately, we cannot require from people to do `--exclude prepare.cri.add_repo` while they do not need to do that now. It can be considered as a some sort of breakage.

> And while the alternative you're talking about is still in the thought experiment phase, this fix will help people who install a kubernetes cluster on nodes with RHEL-like operating systems.

While you fix can work, we cannot include it if is contradicts to the design of the application. By design, installation of repositories happens through the dedicated _task_ and dedicated section of the configuration file. Support of GP code and default public repos may be developed in these task and section for arbitrary repositories in the future. After this is done, it will likely contradict to how you suggest to do that for the particular repository.

Lets' focus on support of RHEL9 family within this PR. This is really valuable for us. The two new features related to the repositories will be handles in the separate issue.

Comment (krang404) [reply kubemarine/cri/containerd.py]: OK. I really understand why you are focused on deploying in secure environments. And I respect your approach to the architecture of the application. So, I have removed adding the docker-ce repository from the installation process.

**Comment by ilia1243** [comment]
Could you resolve some failed jobs?
https://github.com/krang404/KubeMarine/actions/runs/5690554338/job/15424116930
https://github.com/krang404/KubeMarine/actions/runs/5690554338/job/15424116717

File: kubemarine/cri/containerd.py (hunk 1 above)
Comment (ilia1243) [reply kubemarine/cri/containerd.py]: #481

**Comment by koryaga** [comment]
Going to test this PR on internal RHEL9 environment

**Comment by dmyar21** [comment]
Hi @krang404 
Please tell me which image of the system do you use for testing?
Please send the output of the _**cat /etc/*-release**_ command
I want to clarify the id of your image

**Comment by krang404** [comment]
> Hi @krang404 Please tell me which image of the system do you use for testing? Please send the output of the _**cat /etc/*-release**_ command I want to clarify the id of your image

This PR was tested in Oracle Linux 9.2 (1st) and CentOS 9 Stream (2nd). Also, I installed k8s on Rocky Linux 9.2 with subj. 

OL9 
```
NAME="Oracle Linux Server"
VERSION="9.2"
ID="ol"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Oracle Linux Server 9.2"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:oracle:linux:9:2:server"
HOME_URL="https://linux.oracle.com/"
BUG_REPORT_URL="https://github.com/oracle/oracle-linux"

ORACLE_BUGZILLA_PRODUCT="Oracle Linux 9"
ORACLE_BUGZILLA_PRODUCT_VERSION=9.2
ORACLE_SUPPORT_PRODUCT="Oracle Linux"
ORACLE_SUPPORT_PRODUCT_VERSION=9.2
```
CentOS 9 Stream
```
NAME="CentOS Stream"
VERSION="9"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="9"
PLATFORM_ID="platform:el9"
PRETTY_NAME="CentOS Stream 9"
ANSI_COLOR="0;31"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:centos:centos:9"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux 9"
REDHAT_SUPPORT_PRODUCT_VERSION="CentOS Stream"
```
Rocky Linux 9
```
NAME="Rocky Linux"
VERSION="9.2 (Blue Onyx)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Rocky Linux 9.2 (Blue Onyx)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
SUPPORT_END="2032-05-31"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
ROCKY_SUPPORT_PRODUCT_VERSION="9.2"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.2"
```

**Comment by dmyar21** [comment]
> > Hi @krang404 Please tell me which image of the system do you use for testing? Please send the output of the _**cat /etc/*-release**_ command I want to clarify the id of your image
> 
> This PR was tested in Oracle Linux 9.2 (1st) and CentOS 9 Stream (2nd). Also, I installed k8s on Rocky Linux 9.2 with subj.
> 
> OL9
> 
> ```
> NAME="Oracle Linux Server"
> VERSION="9.2"
> ID="ol"
> ID_LIKE="fedora"
> VARIANT="Server"
> VARIANT_ID="server"
> VERSION_ID="9.2"
> PLATFORM_ID="platform:el9"
> PRETTY_NAME="Oracle Linux Server 9.2"
> ANSI_COLOR="0;31"
> CPE_NAME="cpe:/o:oracle:linux:9:2:server"
> HOME_URL="https://linux.oracle.com/"
> BUG_REPORT_URL="https://github.com/oracle/oracle-linux"
> 
> ORACLE_BUGZILLA_PRODUCT="Oracle Linux 9"
> ORACLE_BUGZILLA_PRODUCT_VERSION=9.2
> ORACLE_SUPPORT_PRODUCT="Oracle Linux"
> ORACLE_SUPPORT_PRODUCT_VERSION=9.2
> ```
> 
> CentOS 9 Stream
> 
> ```
> NAME="CentOS Stream"
> VERSION="9"
> ID="centos"
> ID_LIKE="rhel fedora"
> VERSION_ID="9"
> PLATFORM_ID="platform:el9"
> PRETTY_NAME="CentOS Stream 9"
> ANSI_COLOR="0;31"
> LOGO="fedora-logo-icon"
> CPE_NAME="cpe:/o:centos:centos:9"
> HOME_URL="https://centos.org/"
> BUG_REPORT_URL="https://bugzilla.redhat.com/"
> REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux 9"
> REDHAT_SUPPORT_PRODUCT_VERSION="CentOS Stream"
> ```
> 
> Rocky Linux 9
> 
> ```
> NAME="Rocky Linux"
> VERSION="9.2 (Blue Onyx)"
> ID="rocky"
> ID_LIKE="rhel centos fedora"
> VERSION_ID="9.2"
> PLATFORM_ID="platform:el9"
> PRETTY_NAME="Rocky Linux 9.2 (Blue Onyx)"
> ANSI_COLOR="0;32"
> LOGO="fedora-logo-icon"
> CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
> HOME_URL="https://rockylinux.org/"
> BUG_REPORT_URL="https://bugs.rockylinux.org/"
> SUPPORT_END="2032-05-31"
> ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
> ROCKY_SUPPORT_PRODUCT_VERSION="9.2"
> REDHAT_SUPPORT_PRODUCT="Rocky Linux"
> REDHAT_SUPPORT_PRODUCT_VERSION="9.2"
> ```

Hi @krang404 
have you tested RHEL 9 at home?
because I found an interesting point that my RHEL 9 image has an id named rhel.
it turns out your changes with the addition of RHEL 9 to global.yaml not working

> NAME="Red Hat Enterprise Linux"
VERSION="9.2 (Plow)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Red Hat Enterprise Linux 9.2 (Plow)"
ANSI_COLOR="0;31"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:redhat:enterprise_linux:9::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 9"
REDHAT_BUGZILLA_PRODUCT_VERSION=9.2
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.2"
Red Hat Enterprise Linux release 9.2 (Plow)
Red Hat Enterprise Linux release 9.2 (Plow)

**Comment by krang404** [comment]
Red Hat Enterprise Linux version 9 have a problem with installation of `policycoreutils-python-utils` package. And I have not included support RHEL9 in this PR.

**Comment by krang404** [comment]
Also in new versions **iptables** are replaced to package `iptables-nft ` and cluster.yaml will need this section

```
    mandatory:
      iptables: false
    install:
      include:
      - ethtool
      - socat
      - iptables-nft
```

**Comment by dmyar21** [comment]
> Red Hat Enterprise Linux version 9 have a problem with installation of `policycoreutils-python-utils` package. (it is contained in OL9 Appstream Repo, but conflicted with some packages ). So, I am not including support RHEL9 in this PR. Installation was tested for 3 rhle9-like OS.

This package is in the centos repository, you can use it for testing

image

**Comment by krang404** [comment]
> > Red Hat Enterprise Linux version 9 have a problem with installation of `policycoreutils-python-utils` package. (it is contained in OL9 Appstream Repo, but conflicted with some packages ). So, I am not including support RHEL9 in this PR. Installation was tested for 3 rhle9-like OS.
> 
> This package is in the centos repository, you can use it for testing
> image

This package is contained in many repos, but it has many problems with dependencies with python libs in RHEL9

**Comment by dmyar21** [comment]
> > Red Hat Enterprise Linux version 9 have a problem with installation of `policycoreutils-python-utils` package. (it is contained in OL9 Appstream Repo, but conflicted with some packages ). So, I am not including support RHEL9 in this PR. Installation was tested for 3 rhle9-like OS.
> 
> This package is in the centos repository, you can use it for testing

Hi @krang404 
Why then do you add to global.yaml version of rhel9, if you haven't tested it?

**Comment by krang404** [comment]
> 
> Hi @krang404 Why then do you add to global.yaml version of rhel9, if you haven't tested it?

Hello. There is a misunderstanding. `rhel9` in global.yaml means `os_family` and then I created this pr I wrote about rhel9 os family, not Red Hat Enterprise Linux version 9. If you look at the text in the description of PR, you will see it.

**Comment by krang404** [comment]
> 
> Hi @krang404 Please resolve conflicts in your pr

Hello @dmyar21. Done.

**Comment by dmyar21** [comment]
> > Hi @krang404 Please resolve conflicts in your pr
> 
> Hello @dmyar21. Done.

Hi @krang404 
Please update the main branch in your repository from Kubemarine main and make sync fork main to fork selinux-support branch

**Comment by krang404** [comment]
> 
> Hi @krang404 Please update the main branch in your repository from Kubemarine main and make sync fork main to fork selinux-support branch
> 
> Because you have an outdated code there

Done.

**Comment by koryaga** [comment]
@krang404  thanks for efforts. We are planning to adopt this feature in R2023.4. 
If you would like to continue to support this PR pls communicate with @dmyar21 and/or @ilia1243  to update code with latest Kubemarine changes.

**Comment by ilia1243** [comment]
Hi @krang404, please sync main branch with Netcracker/KubeMarine main and merge main to the PR branch.
Alternatively, please invite me as collaborator, I will maintain the branch while we are adopting the feature.

**Comment by ilia1243** [comment]
Hi @krang404, I slightly changed the PR. I left only tested OS and removed Rocky 8.8, OL 9.1. Also changed packages compatibility and changed dealing with SELinux security context of files. Please confirm if the new changes still fit your needs.

**Review by ilia1243 — APPROVED** [review]

**Comment by krang404** [comment]
> Hi @krang404, I slightly changed the PR. I left only tested OS and removed Rocky 8.8, OL 9.1. Also changed packages compatibility and changed dealing with SELinux security context of files. Please confirm if the new changes still fit your needs.
> 
> Also actualized the PR description.

Yes, this is OK, thank you.

**Review by koryaga — APPROVED** [review]

File: documentation/Installation.md [hunk 4]
```diff
@@ -212,7 +212,7 @@ If you have other solution, remove or switch off the IP firewall before the inst
 
 * Installation of the following packages is highly recommended; however, Kubernetes can work without them, but may show warnings:
   * ethtool
-  * ebtables
+  * ebtables (included in the iptables-nft package which is available on systems like rchel 9+)
```
Comment (alexarefev) [inline documentation/Installation.md:215]: May be it's RHEL 9+?

Comment (krang404) [reply documentation/Installation.md]: Edited

**Comment by koryaga** [comment]
@OleksandrIaresko  pls validate in QA and provide feedback

**Comment by koryaga** [comment]
@krang404  can you pls resolve conflicts with main  and add RHEL9 support for newly added k8s 1.28.3 in main as well. After that we will merge this feature

**Comment by krang404** [comment]
Add support to kubernetes v1.28.3 and test it in local environment for OL 9.2 and Rocky 9.2

File: kubemarine/system.py [hunk 5]
```diff
@@ -265,6 +265,10 @@ def patch_systemd_service(group: DeferredGroup, service_name: str, patch_source:
     group.put(io.StringIO(utils.read_internal(patch_source)),
               f"/etc/systemd/system/{service_name}.service.d/{service_name}.conf",
               sudo=True)
+    if group.get_nodes_os() in ['rhel', 'rhel8', 'rhel9']:
+        group.sudo(f"chcon -u system_u -r object_r -t systemd_unit_file_t /etc/systemd/system/{service_name}.service.d")
+        group.sudo(f"chcon -u system_u -r object_r -t systemd_unit_file_t /etc/systemd/system/{service_name}.service.d/{service_name}.conf")
```
Comment (ilia1243) [inline kubemarine/system.py:270]: Hi @krang404 , I see you have lost some my changes. I reworked this explicit chcon invocation to `mv -Z` option when copying the files on nodes. Could you revert the changes and try to rebase again?

Comment (krang404) [reply kubemarine/system.py:270]: Oh, sorry. But if it not changed in my local repo installation in OL 9.2 and Rocky 9.2 are failed. I revert this changes. But are you sure that's ok?

Comment (ilia1243) [reply kubemarine/system.py]: Could you revert all the changes you've made half an hour ago :)
The branch previously contained no `chcon` at all, but contained some other changes to support SELinux

Comment (krang404) [reply kubemarine/system.py]: Now its clean from 'chcon'. And what changes i`ve lost?

Comment (ilia1243) [reply kubemarine/system.py]: All the changes I have done in the branch approximately a week ago. You probably forgot to pull the changes, and discarded them using `--force`. It seems I had to repeat them. Could you add me as collaborator to your project?

Comment (krang404) [reply kubemarine/system.py]: Ops. Shame on me. Invite you to collaboration in my fork. Double sorry.

Comment (ilia1243) [reply kubemarine/system.py]: Updated 5f62bb60ce20b36c6796d012fbd89672d3083f09

**Review by koryaga — APPROVED** [review]
